### ISSUE TITLE
Enhancement: support XSLT3 and XPath3

### ISSUE THREAD

**Issue by rjelliffe** [Schematron/schematron#20]
Schematron skeleton should be upgraded to support XSLT3 and XPath3.  

There are three levels of support:

* The implementation allows a schema/@queryBinding value of  xpath3 and converts this to the appropriate xslt:stylesheet/@version="3.0".   That then enables the new xpath3 syntax.  This can be done immediately, I think.  

* We need to work out a proper Query Language binding for XSLT3.  There are quite a few things in XSLT3 that require thought.  For example,  a big part is type safety.  So I think an XSLT3 binding should support  sch:let/@xsl:as    (in the same way that xsl:key was introduced because it is a query-language-dependent feature.)

* We may need to figure out what the streaming story is for XSLT3.  There is already a STAX QLB available, but would it be useful/necessary to have a query-language-binding subset of for  "xslt3-streaming"

**Comment by PStellmann** [comment]
I would like to get the ability of using sch:let/@xsl:as for XSLT2 binding as well.

My motivation is that in XSLT I'm used to add an @as attribute for every xsl:param and xsk:variable since I consider this to be cleaner code . So I'd like to do the same for my schematron rules.

**Comment by rjelliffe** [comment]
Sounds reasonable.  It would not fit in with the ISO standard's query
language binding for XSLT2, but that is not reason not to do it or provide
it.

Regards
Rick

**Comment by DemiMarie** [comment]
One difficulty is that I don't know of *any* open source implementations of XSLT3.

**Comment by rjelliffe** [comment]
Yes. On XML-DEV mainlist I asked Michael Kay of SAXON fame whether the
strong typing did in fact allow any optimization, and the impression I got
was that it didn't really (though it could in the future... hmmm)

Furthermore, some typing-checking should really be done with assertions.  I
wonder whether we need assertions on variables?

So really that leaves documentation and XPath checking as the reasons for
it, which was what Patrik was asking for.  I suppose the easiest way is to
just define a foreign namespace and allow for example  sch:let/@type:as
which could be declared with xpath1, xpath2, xpath3 but only used for
implementations that support strong typing.

Rick

On Thu, Mar 2, 2017 at 8:06 AM, Demi Marie Obenour <notifications@github.com

**Comment by nigelwhitaker** [comment]
I've been working on a solution to the first level of support (first bullet in the first comment from Rick) and am hoping to create a pull request when I've run existing tests and added some new tests.
